include/tune_quda.h

```diff
@@ -428,18 +412,28 @@ namespace quda {
      The x threads will typically correspond to the checkboarded
      volume.
    */
-  class TunableLocalParity : public Tunable {
+  class TunableLocalParityReduction : public Tunable
+  {
 
   protected:
     unsigned int sharedBytesPerThread() const { return 0; }
     unsigned int sharedBytesPerBlock(const TuneParam &param) const { return 0; }
 
-    // don't tune the grid dimension
-    virtual bool tuneGridDim() const { return false; }
+    /**
+       Reduction kernels require grid-size tuning, so enable this and
+       make it non-virtual so that if user accidentally disables it,
+       compiler will complain
+    */
+    bool tuneGridDim() const { return true; }
```

Comment:
(I could be wrong here - this is based on my remote memory of how virtual functions work) I don't think dropping the `virtual` specifier here would make `tuneGridDim()` non-virtual: as long as `tuneGridDim()` is virtual in the base class `Tunable`, all `tuneGridDim()` in the derived classes are all considered virtual. I tried setting `tuneGridDim()` to return false in this [class,](https://github.com/lattice/quda/blob/8e323890793d36fd5a8f8887f9f84bffbbc72aa3/lib/contract.cu#L94) which inherits from `TunableLocalParityReduction`, and there is no compiler warning given.

Also in fact for the contraction code we do want to disable `gridDim` tuning since there `gridDim.z` is the number of time slices so we don't want to tune it. Actually in that case the `advanceBlockDim` here resets `gridDim.z` to `1` when ever `Tunable::advanceBlockDim` fails, which is not the intended behavior.

Reply:
You have good points here.
* I'll revert `tuneGridDim` to being virtual and remove the comment.
* Regarding `Tunable::advanceBlockDim` reseting `gridDim.y` and `gridDim.z`, this is actually a bug.  It's something that I fixed in Dean's ongoing experimental branch, and forgot to port to mainline QUDA.  I'll also add the fix for that in my next push.

Reply:
https://github.com/lattice/quda/pull/1046/commits/baff1c84ce93c925ce31470dd4d58b2471ceee17